### PR TITLE
Event Cart cleanup

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
**Please review and merge https://github.com/civicrm/civicrm-core/pull/16708 first.**

Event cart code is a mess and very old. It doesn't interact properly with lot's of bits of CiviCRM (eg. payments).

Before
----------------------------------------
Unmaintainable code.

After
----------------------------------------
Slightly more maintainable code.

Technical Details
----------------------------------------
- Code cleanup in event cart: General comments, typos, cleanups.
- Remove non working code that makes event cart behave differently if you are logged in as administrator: Slightly different/extra fields shown, call to a update_note function that does not exist anymore etc.
- Handle non-default currency: Event cart is very opinionated about using the default currency and not passing the actual currency as a property on the form (as all other forms do).
- Cleanup 'View Cart' form markup: Add a "no items in cart" text, wrapper div and whitespace.
- Split event list page from ical/rss feeds: CRM/Event/Page/ICalendar is a nasty mix of things which outputs a page, a rss feed, a gcal feed and an ical feed. Separate the page and the feeds code - see https://github.com/civicrm/civicrm-core/pull/16708
- Cleanup ParticipantsAndPrices template for event cart: Code cleanup / formatting

Comments
----------------------------------------
This is broken down into separate commits for each bit of cleanup so you can see more clearly what is being changed.

### CODE REVIEW

File: CRM/Event/Cart/Form/Checkout/Payment.php [hunk 1]
```diff
@@ -25,22 +30,16 @@ class CRM_Event_Cart_Form_Checkout_Payment extends CRM_Event_Cart_Form_Cart {
...
   public function registerParticipant($params, &$participant, $event) {
-    $transaction = new CRM_Core_Transaction();
-
-    // handle register date CRM-4320
-    $registerDate = date('YmdHis');
     $participantParams = [
       'id' => $participant->id,
       'event_id' => $event->id,
-      'register_date' => $registerDate,
+      'register_date' => date('YmdHis'),
```
[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply CRM/Event/Cart/Form/Checkout/Payment.php]: I'm pretty sure both work. I could make a case  for the BAO having a default & it being optional

[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply CRM/Event/Cart/Form/Checkout/Payment.php]: But even if both work - whatever is more commonly used is better

[the remark this answers is not in the document]
Comment (mattwire) [reply CRM/Event/Cart/Form/Checkout/Payment.php]: So this is really a no change - as we're just eliminating a variable that is only used in one place and moving the definition. @eileenmcnaughton @seamuslee001 Can we mark this as resolved and, if necessary, follow up. There is lot's of scope for moving various calls to use the API within the event cart code etc.

File: CRM/Event/xml/Menu/Event.xml [hunk 2]
```diff
@@ -44,11 +44,19 @@
   <item>
      <path>civicrm/event/ical</path>
      <title>Current and Upcoming Events</title>
-     <page_callback>CRM_Event_Page_ICalendar</page_callback>
+     <page_callback>CRM_Event_ICalendar::run</page_callback>
```
[the remark this answers is not in the document]
Comment (mattwire) [reply CRM/Event/xml/Menu/Event.xml]: Yes no problem - wanted to put the changes up first to see if tests pass etc and what the general feeling was - separate PR here: https://github.com/civicrm/civicrm-core/pull/16708

File: CRM/Event/Cart/Form/MerParticipant.php [hunk 3]
```diff
@@ -2,30 +2,31 @@
 
 /**
  * Class CRM_Event_Cart_Form_MerParticipant
+ * @fixme What is a MerParticipant!
```
Comment (colemanw) [inline CRM/Event/Cart/Form/MerParticipant.php]: MerParticipants?
MerParticipants

**Comment by colemanw** [comment]
Code looks good to me. @mlutfy you're the only one I know of who actually uses this feature would you be able to give it a spin?
Once this is merged I would suggest a "phase II" cleanup would be moving Event Cart into a core extension (something recently invented by @eileenmcnaughton)

**Comment by mattwire** [comment]
> Once this is merged I would suggest a "phase II" cleanup would be moving Event Cart into a core extension
@colemanw Yes, I'm very much hoping we can move towards that.

File: CRM/Event/Cart/Form/MerParticipant.php (hunk 3 above)
Comment (mattwire) [reply CRM/Event/Cart/Form/MerParticipant.php]: @colemanw Ah, thanks... mystery solved :-)

**Comment by mlutfy** [comment]
Seems good to me. Thanks @mattwire!

I think a MerParticipant is a participant in a Cart.

It's documented here:  
https://github.com/civicrm/civicrm-core/blob/master/CRM/Event/Cart/Form/MerParticipant.php#L111

;-)

**Comment by demeritcowboy** [comment]
You know how sometimes there's something you can't let go without figuring it out. I'm pretty sure now it's "Multiple Event Registration Participant". See also https://issues.civicrm.org/jira/browse/CRM-8705 which includes a "MerPatch".

**Comment by eileenmcnaughton** [comment]
@demeritcowboy  - for when your participants have fishy tails

**Comment by eileenmcnaughton** [comment]
ah - I see the gif now - I'm well too late on that

**Pull request merged**